### PR TITLE
dev/core#2121 Ability to change pdf filename before downloading

### PR DESCRIPTION
**Pull request by deb1990**
Overview
----------------------------------------
This is a continuation of https://github.com/civicrm/civicrm-core/pull/18938.

For Print/Merge Activities, the pdf filename was always downloaded with the name `CiviLetter.pdf`. So in this PR, we are allowing the filename to be customised.

By default the Subject of the Activity will be considered as the file name.
If No Subject is available, then existing `CiviLetter` will still be used.
Also any 3rd party extensions can customise the name even further by submitting the form with `pdf_file_name` hidden field added.

Before
----------------------------------------
The filename was always CiviLetter.pdf.

After
----------------------------------------
98368463-30edc880-2059-11eb-96b3-b0d6947e585f

Technical Details
----------------------------------------
In `CRM/Core/Form/Task/PDFLetterCommon.php`, a new hidden field `pdf_file_name` is added, using which the pdf filename can be customised by any 3rd party extensions. 

In `CRM/Contact/Form/Task/PDFLetterCommon.php`, added logic to apply the `pdf_file_name` filed values as pdf file name, if its not present then Activity Subject will be considered, and if that is also not present, the previous `CiviLetter` string will be used as PDF file name.

### CODE REVIEW

File: CRM/Contact/Form/Task/PDFLetterCommon.php [hunk 1]
```diff
@@ -215,6 +216,29 @@ public static function postProcess(&$form) {
...
+  /**
+   * Returns the filename for the pdf by striping off unwanted characters and limits the length to 200 characters.
+   *
+   * @param CRM_Core_Form $form
+   *
+   * @return string
+   *   The name of the file.
+   */
+  private static function getFileName(CRM_Core_Form $form) {
+    if (!empty($form->getVar('_submitValues')['pdf_file_name'])) {
```
Comment (eileenmcnaughton) [inline CRM/Contact/Form/Task/PDFLetterCommon.php:228]: @monishdeb this line was specifically raised as a concern in the previous PR

Comment (eileenmcnaughton) [reply CRM/Contact/Form/Task/PDFLetterCommon.php:228]: sorry - it wasn't Monish it was @deb1990

**Comment by eileenmcnaughton** [comment]
@mlutfy - I tried re-reading https://lab.civicrm.org/dev/core/-/issues/2121 to figure out how this fit with what was agreed but it will take more than a light skim to make sense of - you originally had concerns - maybe you could re-parse this?

File: CRM/Contact/Form/Task/PDFLetterCommon.php (hunk 1 above)
Comment (deb1990) [reply CRM/Contact/Form/Task/PDFLetterCommon.php]: @eileenmcnaughton it is fixed now.

Comment (eileenmcnaughton) [reply CRM/Contact/Form/Task/PDFLetterCommon.php]: Yay - looks better

File: CRM/Core/Form.php [hunk 2]
```diff
@@ -2653,6 +2653,25 @@ public function getCurrency($submittedValues = []) {
...
+  /**
+   * Get the submitted value of a form field.
+   *
+   * @param array $fieldName
+   *   Form field name.
+   *
+   * @return string
+   *   Submitted value of the form field.
+   */
+  public function getSubmittedValue ($fieldName) {
```
Comment (seamuslee001) [inline CRM/Core/Form.php:2665]: Why is there a space here between function name and (

Comment (deb1990) [reply CRM/Core/Form.php]: Fixed.

**Comment by mlutfy** [comment]
@eileenmcnaughton Using the activity subject makes sense to me, and addresses my concern of adding extra clicks/questions.

**Comment by seamuslee001** [comment]
@deb1990 error from test site is 

```
PHP Fatal error:  Cannot redeclare CRM_Core_Form::getSubmittedValue() in CRM/Core/Form.php on line 2766

Fatal error: Cannot redeclare CRM_Core_Form::getSubmittedValue() in CRM/Core/Form.php on line 2766
```

**Comment by deb1990** [comment]
@eileenmcnaughton @seamuslee001 Fixed it now, also I had to convert the existing `getSubmittedValue` function to public in order to use it. So let me know if this looks good now.

**Comment by demeritcowboy** [comment]
On the `r-explain` I think it's important to note that the issue is not so much the name of the file downloaded to your computer. It's that if you have changed the setting at Administer - System Settings - Misc for "Record generated letters" and set it to combined+attachment, then the name of the attachment that gets stored on the activity is always CiviLetter.pdf and can't be changed in the UI. It mentions this is the point in the lab ticket but doesn't mention the system setting.

**Comment by eileenmcnaughton** [comment]
I have my normal interface concerns here - there is an intent that we add a hidden field in order for extensions to be able to alter the value. In general we have 2 ways extensions can interact with core

1) supported tested documented ways which extensions can expect to continue to work
2) unsupported ways which could break & which would not be core's problem if they did

This seems to be adding a new case of 2 - but with a comment implying 1 within the code and code within core specifically to support it. I guess tests & documentation are the minimum required to move from 2 to 1.

An alternative would be to populate the hidden field with javascript and possibly (I can't believe I'm about to suggest this) have a setting to determine whether or not it is hidden

**Comment by eileenmcnaughton** [comment]
I guess just to repeat what I see are the competing concerns

1) @deb1990 wants to be able to alter the filename from an extension
2) I don't want us to have an 'implied' support expectation that is not locked in by tests / docs
3) @mlutfy doesn't want to clutter the UI with another box

Does that represent the concerns correctly?

**Comment by demeritcowboy** [comment]
How is the extension proposing to populate it? I was assuming the buildForm hook. But yes it might be a little difficult to write a test for that here, but might be technically possible.

A new hook here instead might work and be easier to test, but it would be a very very specific hook: `hook_civicrm_alterPrintMergeDocumentAttachmentFilename`

I'm not sure I understand the javascript bit - I mean I understand you can populate the field with javascript - but not seeing how it addresses "supported tested documented ways which extensions can expect to continue to work"?

**Comment by eileenmcnaughton** [comment]
@demeritcowboy yeah I guess the js theory puts it 'at the same level of supportedness as any other field'

I did wonder if @colemanw could think about the UI side of this - I mean it seems like it could be an exposed field if we had a way to do it in the UI that wasn't insanely ugly.

Also noting that @deb1990 is ready to write the tests for the scenario listed below - which would address my concerns - I just hoped to get a little bit of @colemanw's UI attention before saying that is the best approach.


When pdf_file_name value is submitted, the filename is generated using the same.
When pdf_file_name value is NOT submitted, the "Activity Subject" field is used to generate the filename,
When non of them are present, CiviLetterstring is used as the filename.

**Comment by deb1990** [comment]
@demeritcowboy It can be populated using JS before submitting the form, 
`$('[name="pdf_file_name"]').val('<VALUE FOR THE FILE NAME>');`

**Comment by eileenmcnaughton** [comment]
@deb1990 you are right - I guess even if we decide to unhide the fields your tests still make sense - go ahead & write them & we can merge I think. I'm on the fence about the importance of documenting - it's the tests that make in maintainable & I guess it's as discoverable by a dev as any other field (ahem) so I think that would be very much optional

**Comment by agileware-justin** [comment]
This would be a useful change.

**Comment by eileenmcnaughton** [comment]
@agileware-justin looks like you spurred some action!

Looking through this discussion I'm not clear that someone has given this an r-run but if you want to do that I think we can merge this

**Comment by deb1990** [comment]
@eileenmcnaughton Actually we are in the process of writing the unit test cases. Should be ready in couple of days. Apologies for the delay, again.

**Comment by lisandro-compucorp** [comment]
hello @eileenmcnaughton, I added the tests, as discussed with @deb1990 . I did not ping you before for checking it because I was waiting for the checks to finish, and it was late in my timezone. Let me know if that is ok. Thanks!

**Comment by eileenmcnaughton** [comment]
@lisandro-compucorp yes - tests looks good - this just needs a reviewer to do a quick r-run & we can merge

**Comment by agileware-justin** [comment]
> @agileware-justin looks like you spurred some action!
> 
> Looking through this discussion I'm not clear that someone has given this an r-run but if you want to do that I think we can merge this

@eileenmcnaughton happy to test this PR tomorrow

**Comment by agileware-justin** [comment]
Testing now. Unfortunately, my Bees needed a lot more work than I expected yesterday - so didn't have any spare time.

**Comment by agileware-justin** [comment]
@deb1990 are you able to update this PR to include the same logic applied to the Membership PDF. The Activity Subject is available on the form. See https://github.com/civicrm/civicrm-core/blob/master/CRM/Member/Form/Task/PDFLetterCommon.php#L36

This would address the "CiviLetter.pdf" issue generated for the **Thank-you Letter for Contributions (PDF)** as shown below.

Screenshot_20210823_103634

**Comment by eileenmcnaughton** [comment]
@agileware-justin is that request blocking on merging this?

**Comment by agileware-justin** [comment]
Can confirm that for Print/Merge Activities, the PDF file name is the Activity Subject, if provided. Instead of always being "CiviLetter.pdf"

Screenshot_20210823_103901

**Comment by agileware-justin** [comment]
> @agileware-justin is that request blocking on merging this?

No, it's not blocking. However, for consistency this needs at least one follow-up PR (and possibly more) to address the other parts of CiviCRM which are still outputting "CiviLetter.pdf" when performing the same type of action, ie. creating a PDF from a Message Template and having the Activity Subject displayed.

**Comment by agileware-justin** [comment]
OK to merge. More PRs to follow.

**Comment by agileware-justin** [comment]
Here's another example, "CiviContribute.pdf"

Screenshot_20210823_103959

File: CRM/Contact/Form/Task/PDFLetterCommon.php [hunk 3]
```diff
@@ -215,6 +216,29 @@ public static function postProcess(&$form) {
...
+   */
+  private static function getFileName(CRM_Core_Form $form) {
+    if (!empty($form->getSubmittedValue('pdf_file_name'))) {
+      $fileName = CRM_Utils_String::munge($form->getSubmittedValue('pdf_file_name'), '_', 200);
+    }
+    elseif (!empty($form->getSubmittedValue('subject'))) {
+      $fileName = CRM_Utils_String::munge($form->getSubmittedValue('subject'), '_', 200);
+    }
+    else {
+      $fileName = "CiviLetter";
```
Comment (agileware-justin) [inline CRM/Contact/Form/Task/PDFLetterCommon.php:235]: Ideally, this should be  'CiviLetter';

No need for double-quotes here.

File: CRM/Contact/Form/Task/PDFLetterCommon.php [hunk 4]
```diff
@@ -215,6 +216,29 @@ public static function postProcess(&$form) {
...
+    }
+    $fileName = self::isLiveMode($form) ? $fileName : $fileName . "_preview";
```
Comment (agileware-justin) [inline CRM/Contact/Form/Task/PDFLetterCommon.php:237]: Same applies here, '_preview';

No need for double-quotes here.

**Comment by agileware-justin** [comment]
Just noting the follow-up PR is here, https://github.com/civicrm/civicrm-core/pull/21220

**Comment by deb1990** [comment]
Thanks @agileware-justin for the review and the PR. Fixed the same.

**Comment by eileenmcnaughton** [comment]
@agileware-justin I'm happy to merge this on your say-so

**Comment by deb1990** [comment]
@agileware-justin Please confirm if the PR looks to you :)

**Comment by deb1990** [comment]
@eileenmcnaughton We already have a thumbs up from @agileware-justin in my last comment :), so maybe we can merge this?

**Comment by eileenmcnaughton** [comment]
ok

**Comment by agileware-justin** [comment]
Thumbs up

**Comment by agileware-justin** [comment]
:smile:

**Comment by eileenmcnaughton** [comment]
Not THAT is a convincing thumbs up

**Comment by deb1990** [comment]
Hahaha, thanks everyone from reviewing the PR.

**Comment by jamienovick** [comment]
A BIG thank you to everyone for their time on this - great to get it closed off :)

**Pull request merged**